File: 0055-jump-game/0055-jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        maxReach = nums[0]
        for i in range(len(nums)):
            maxReach = max(maxReach, i + nums[i])
            if maxReach >= len(nums) - 1:
                return True
            if nums[i] == 0 and maxReach == i:
                return False            
        return False

File: 0055-jump-game/0055-jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        int[] dp = new int[nums.length];
        return solve(nums, 0, dp);
    }

    boolean solve(int[] nums, int start, int[] dp){
        if (start >= nums.length - 1){
            return true;
        }
        if (dp[start] != 0){
            return dp[start] == 1 ? true : false;
        }
        if (nums[start] == 0){
            return false;
        }
        for(int i=1; i<=nums[start]; i++){
            if (solve(nums, start + i, dp) == true){
                dp[start] = 1;
                return true;
            }
        }
        dp[start] = -1;
        return false;
    }
}